File: cpp/symmetric_tree.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isSymmetric(TreeNode *root) {
        if (!root)
            return true;
        stack<TreeNode *> node_stack;
        TreeNode *left, *right;
        node_stack.push(root->left);
        node_stack.push(root->right);
        while (!node_stack.empty())
        {
            right = node_stack.top();
            node_stack.pop();
            left = node_stack.top();
            node_stack.pop();
            if (!left && !right)
                continue;
            else if (!left || !right)
                return false;
            else if (left->val != right->val)
                return false;
            else
            {
                node_stack.push(left->left);
                node_stack.push(right->right);
                node_stack.push(left->right);
                node_stack.push(right->left);
            }
        }
        return true;
    }
};

File: python/symmetric_tree.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return a boolean
    def isSymmetric(self, root):
        if root == None:
            return True
        else:
            return self.isSymmetric2(root.left, root.right)
    
    def isSymmetric2(self, left, right):
        if left == None and right == None:
            return True
        elif left == None or right == None:
            return False
        elif left.val != right.val:
            return False
        else:
            return self.isSymmetric2(left.left, right.right) and self.isSymmetric2(left.right, right.left)